go/host/enclave/service.go

```diff
@@ -200,27 +175,108 @@ func (e *Service) Unsubscribe(id rpc.ID) error {
+				// todo: do we want to have a few failed attempts before looking to promote a new one?
+				e.logger.Error("Sequencer failed to produce batch", log.ErrKey, err)
+				// if the active sequencer fails, we will try to promote a new one
+				e.tryPromoteNewSequencer()
 				continue
 			}
-			e.activeSequencerID.Store(enclID)
-			e.logger.Warn("Successfully promoted new sequencer.", log.EnclaveIDKey, enclID)
-			return
 		}
-		// wait for retry interval before trying again, enclaves may not be ready yet or may be awaiting permissioning
-		time.Sleep(_promoteSeqRetryInterval)
+
+		// sanity check: make sure there are no enclaves, which think they are active sequencers
+		for _, guardian := range e.enclaveGuardians {
+			if guardian.state.IsEnclaveActiveSequencer() && guardian != activeSeq {
+				e.logger.Warn("Found unexpected active sequencer in guardians, demoting...",
+					log.EnclaveIDKey, guardian.GetEnclaveID(), "expectedActiveID", activeSeq.GetEnclaveID().Hex())
+				guardian.DemoteFromActiveSequencer()
+			}
+		}
 	}
+	batchProductionTicker.Stop()
+	e.logger.Info("Stopping periodic batch production.")
 }
 
+func (e *Service) tryPromoteNewSequencer() {
+	prevActiveSeqID := e.activeSequencerID.Load()
+
+	// prepare a list of guardians to try, starting with the one that thinks it is active (if there is one, e.g. after host restart)
+	// and skipping the one that just got demoted (if there was one)
+	guardiansToTry := make([]*Guardian, 0, len(e.enclaveGuardians))
+
+	for _, guardian := range e.enclaveGuardians {
+		enclID := guardian.GetEnclaveID()
+		if prevActiveSeqID != nil && *enclID == *prevActiveSeqID {
+			// skip this enclave since we are replacing it
+			continue
+		}
+
+		// if the network is just starting up the candidate just needs to be in sync with L1
+		networkStartingUp := e.sl.L2Repo().FetchLatestBatchSeqNo().Cmp(big.NewInt(1)) < 0
+		if networkStartingUp && guardian.InSyncWithL1() || guardian.IsLive() {
+			if guardian.state.IsEnclaveActiveSequencer() {
+				// prepend this guardian to the list of guardians to try because it was active
+				guardiansToTry = append([]*Guardian{guardian}, guardiansToTry...)
+			} else {
+				guardiansToTry = append(guardiansToTry, guardian)
+			}
+		}
+	}
+
+	// attempt to promote one of the candidate guardians
+	for _, guardian := range guardiansToTry {
+		enclID := guardian.GetEnclaveID()
+		e.logger.Debug("Attempting to promote enclave to active sequencer", log.EnclaveIDKey, enclID)
+		err := guardian.PromoteToActiveSequencer()
+		if err != nil {
+			e.logger.Error("Failed to promote enclave to active sequencer", log.ErrKey, err)
+			continue
+		}
+		prevActiveGuardian, err := e.getActiveSequencerGuardian()
+		if err == nil {
+			// if we have a previous active guardian, demote it now (makes sure the enclave is stopped/restarted)
+			e.logger.Info("Demoting previous active sequencer", log.EnclaveIDKey, prevActiveGuardian.GetEnclaveID())
+			prevActiveGuardian.DemoteFromActiveSequencer()
```

Comment:
shouldn't the order be: 1. demote, 2. promote ?
Probably not an issue here because it's a single thread, but who knows

Reply:
I think it's important we don't demote (which will shutdown the enclave) unless we've already promoted successfully to avoid weird situations where we end up with no active sequencer. If both enclaves have issues then we might as well wait for the current active sequencer to recover. But yeah because we drive the batch production from this single thread now there's no risk of double-sequencer issues from this.

Reply:
Cool. Let's add a comment to revisit once we make the demoting smarter.